### PR TITLE
Add accountID to the set password link

### PR DESCRIPTION
**Pull request by jasperhuangg**
### Details
The `SetPassword` API command requires an email to be provided, along with a new password and a validateCode. For the `SetPasswordPage`, we were previously assuming that the email would be read from Onyx storage. 

However, that may not be the case if we send the user a set password link via a notification, as the user may not have completed the first part of the sign-in flow (which saves their email into Onyx storage). If this is the case, the API call to `SetPassword` will fail since the email sent over will be `undefined`.

In lieu of the changes in https://github.com/Expensify/Web-Expensify/pull/30466/ and https://github.com/Expensify/Auth/pull/5402, I'm adding an email parameter to the set password route to ensure we're always sending over and reading an email when the user navigates from this page.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/156786
$ https://github.com/Expensify/Expensify/issues/157873

### Tests (local dev)
1. On Web-Expensify, check out `jasper-ecashValidatePasswordEmail`; on Auth, check out and make `jasper-resendValidateCodeSetCashPassword`.
2. Make sure you're following the steps with an unvalidated Expensify email. Run `./script/sql.sh` then:
- Get your accountID:
```
SELECT validated from accounts WHERE email = "[your email]@expensify.com";
```
- Make sure your account is invalidated with this command:
```
UPDATE logins SET validatedDate = "" WHERE accountID = [your accountID];
``` 
- Make sure you've got the most up to date partnerName for Expensify.cash:
```
INSERT OR IGNORE INTO partners ( created, partnerID, partnerName, partnerPassword, partnerKey ) VALUES ( '2021-03-18 00:00:00', 84, 'chat-expensify-com', 'DD1431EB9E11F0071EC05C24E9CB2FD595B1ED75', 'asdf,fdsa');
```
3. Hit `expensify.com.dev/api?command=ResendValidateCode&referer=https://expensify.cash/&accountID=[your accountID]` to queue up a job that will send yourself a set password email.
- Use `SELECT * from notifications;` and check to make sure there's a `SetCashPassword` notification with your accountID queued up in the last row.
4. Then run `vssh` and `php /vagrant/Web-Expensify/script/notifyall.php` to send it out.
- to troubleshoot bedrock jobs, use the [dev Bedrock Job Manager](https://www.expensify.com.dev/_devportal/tools/bedrockJobManager/)
5. Check your email for an email with the subject "Your magic sign in link for Expensify.cash". Click on the link.
6. Modify the first part of the URL with whatever localhost:[port] your local version of cash is running on (e.g. localhost:8081)
- After this, the URL should look something like this: localhost:8081/setpassword/[accountID]/[validateCode]
7. Navigate to that new URL and set a new password (it's better to do this in an incognito window).
8. Sign out of e.cash and try to sign in to that email with the new password. Verify that you can sign in.

Screen Shot 2021-04-07 at 9 44 39 AM

### QA Test Steps
1. Navigate to the sign in page.
2. Create a new account with an email address you have access to (I recommend using [EmailOnDeck](https://www.emailondeck.com))
3. Verify that an email (like the screenshot) was sent to that email address.
4. Click "Resend Link". Verify that another email was sent to that email address.
5. In the newer email, click `here`. Verify that it takes you to `https://expensify.cash/setpassword/:accountID/:validateCode`
6. Set a password, it should log you in after you click "Set Password"
7. Sign out. Verify that you can sign back in with the same email address and password you just set.
- NOTE: this can also be tested on iOS/Android

### Tested On

- [x] Web
- [x] Mobile Web
- [ ] Desktop
- [x] iOS
- [x] Android

#### Web
Screen Shot 2021-03-18 at 11 19 24 AM

#### Mobile Web
Screen Shot 2021-03-18 at 11 18 40 AM

### CODE REVIEW

**Comment by shawnborton** [comment]
For your mobile web screenshot, is there a way we can give the whole page a background color so that we don't have the strange background color block below the form content?

**Comment by jasperhuangg** [comment]
@shawnborton For sure! I made a [new issue](https://github.com/Expensify/Expensify/issues/157816) since it's not entirely related to what I'm fixing here. I've included more details in the issue. Look out for a PR soon!

**Comment by marcaaron** [comment]
>QUESTION FOR REVIEWER: Is this testable on Desktop/iOS/Android?

**Short answer:** No.
**Longer answer:** Once we have "deep links" set up for iOS and Android we can theoretically drop you straight into the app on your device and everything should work the same (see [Deep Linking in react-nav docs](https://reactnavigation.org/docs/deep-linking/)). There is an issue [here](https://github.com/Expensify/Expensify/issues/146908) to address this. As for Desktop, I think it just wouldn't be very valuable to implement right now, but maybe in the future.

File: src/libs/actions/Session.js
Comment (marcaaron) [inline src/libs/actions/Session.js:213]: This probably can be non-optional as we are only already defaulting to an empty string when calling this method.

Also NAB since I think we should change this, but optional params are documented with brackets like this...

```js
 * @param {String} [email]
```

File: src/libs/actions/Session.js
Comment (marcaaron) [inline src/libs/actions/Session.js:223]: Curious why we are changing this from `email` to `login`?
What response params does the API for `SetPassword` send back?
Are we modifying them in the Web-E PR?

**Review by marcaaron — CHANGES_REQUESTED** [review]
Just have a couple questions so far.

**Comment by AndrewGable** [comment]
I'm curious if this issue is related: https://github.com/Expensify/Expensify/issues/157873  to this PR?

**Comment by marcaaron** [comment]
Hmm not sure if related.. the error you are seeing there seems to suggest that we have not set the login in credentials.

https://github.com/Expensify/Expensify.cash/blob/ec2ffc28f0d988cb73daad44c037534ba676c287/src/libs/actions/Session.js#L216

File: src/libs/actions/Session.js
Comment (jasperhuangg) [reply src/libs/actions/Session.js:223]: @marcaaron There's another variable called `email` in the upper scope, so `eslint` points it out as an error.

Comment (marcaaron) [reply src/libs/actions/Session.js:223]: Makes sense. So, is there a `login` property on the `response`? 
If so then we can [destructure](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#object_destructuring). If not, `login` will be `undefined`.

Comment (jasperhuangg) [reply src/libs/actions/Session.js:223]: Ahh you'd definitely be correct in saying that it would be `undefined`. Auth actually sends back the email field, that's totally my bad! I can refrain from using object destructuring here which should solve the problem.

File: src/libs/API.js
Comment (marcaaron) [inline src/libs/API.js:537]: Are we changing the API method name too since we are changing the `commandName`? 🤔

File: src/libs/actions/Session.js
Comment (marcaaron) [inline src/libs/actions/Session.js:224]: NAB, we can just use these directly in `createTemporaryLogin()` instead of creating variables for them.

File: src/libs/API.js
Comment (jasperhuangg) [reply src/libs/API.js:537]: Yes! I'm actually going to change it back to ResendValidateCode today, and modify `ResendValidateCode` in Auth to send out the right notification.
```suggestion
    const commandName = 'ResendValidateCode';
```

File: src/ROUTES.js
Comment (chiragsalian) [inline src/ROUTES.js:20]: I would much rather we pass accountID here instead of email. This way we don't have to deal with special characters in the email breaking our code.

File: src/libs/actions/Session.js
Comment (chiragsalian) [inline src/libs/actions/Session.js:219]: wrt my previous comment lets pass accountID as an additional param then the PHP code for setPassword will extract email from the account ID if provided to make this a lot more simpler.

**Review by chiragsalian — CHANGES_REQUESTED** [review]

File: src/libs/API.js
Comment (chiragsalian) [inline src/libs/API.js:548]: you might want to update the params description of this method to capture parameters.accountID

**Review by chiragsalian — CHANGES_REQUESTED** [review]

**Comment by marcaaron** [comment]
LGTM (besides everything @chiragsalian mentioned so far)

File: src/libs/actions/Session.js
Comment (jasperhuangg) [reply src/libs/actions/Session.js:214]: I figured it would be good to leave in email since the API endpoint takes in either email or accountID. May be good for future use, but if it isn't going to be used anywhere else it's probably better just to get rid of it!

**Comment by chiragsalian** [comment]
Is this ready for review? If so you'll need to let us know by posting "Ready for review" and / or clicking the re-request review icon next to our names in the reviewers section.

**Comment by jasperhuangg** [comment]
@chiragsalian I left a comment on your review that I thought would be better for you to respond to first before reviewing, but looking back I should have just requested another review. Thanks for making sure I stay on top of this and sorry for all the back and forth!

**Review by chiragsalian — CHANGES_REQUESTED** [review]

**Comment by jasperhuangg** [comment]
> > I left a comment on your review that I thought would be better for you to respond to first before reviewing
> 
> Just so I'm on the same page are you talking about a comment on my review in this PR or elsewhere that you are waiting for a response? Either way, if i haven't responded to it can you link me which comment you're waiting for a response from 🙂
> 
> Also yeah no worries about the back and forth. As long as there is daily movement its all cool.

Looks like you were able to get to my comment anyways! Thanks :)

**Review by chiragsalian — DISMISSED** [review]
LGTM, all yours @marcaaron

**Comment by marcaaron** [comment]
Is this still on HOLD ?

**Comment by chiragsalian** [comment]
@marcaaron, Yes because https://github.com/Expensify/Web-Expensify/pull/30466 isn't live yet.

File: src/ROUTES.js
Comment (marcaaron) [inline src/ROUTES.js:23]: These changes don't look like they should be here.

```js
    PROFILE: 'profile',
    PROFILE_WITH_LOGIN: 'profile/:login',
    getProfileRoute: login => `profile/${login}`,
```

I think `profile` was changed to `details` recently so that's not super obvious but will need to be fixed.

File: src/ROUTES.js
Comment (marcaaron) [inline src/ROUTES.js:20]: I'm not sure we need to do anything about this. But it would be good to start thinking about how we would handle such a change in the future. 

e.g. we are breaking any existing links sitting in people's inboxes since the new app will treat `validateCode` as `accountID` and if they get a new deep link but still have the old app then we will have the opposite problem.

**Review by marcaaron — DISMISSED** [review]
Changes LGTM. Leaving this unmerged as I'm not sure when it would be deployed.

File: src/ROUTES.js
Comment (marcaaron) [reply src/ROUTES.js:20]: I think it would be a good idea to create some kind of issue to discuss a solution for problems like this just to get it on people's radar.

**Review by marcaaron — DISMISSED** [review]

File: src/libs/API.js
Comment (tgolen) [inline src/libs/API.js:551]: There are some missing quotes here:
```suggestion
    requireParameters(['accountID', 'password', 'validateCode'], parameters, commandName);
```

File: src/pages/SetPasswordPage.js
Comment (tgolen) [inline src/pages/SetPasswordPage.js:40]: These are all missing inline comments to explain them, can you please add them?

**Review by tgolen — DISMISSED** [review]
I'm a little confused about the title of this PR. Isn't this adding an `accountID` to the set password link? The title says it's adding `email` to the link.

**Comment by marcaaron** [comment]
@jasperhuangg let's add some QA test steps to the description in addition to the local dev steps?

File: src/libs/API.js
Comment (jasperhuangg) [reply src/libs/API.js:551]: Ahh you're right! This was before when I was still adding the email (we changed it to use accountID so that it's more consistent with E.com and URL safe). I've changed the title of the PR, sorry for the confusion!

**Review by marcaaron — APPROVED** [review]
This LGTM. I tested on iOS and also Android since I noticed we did not test on those platforms. We should let QA know that they can test iOS and Android by clicking on the links with the app installed (at least this should work in theory).

**Comment by jasperhuangg** [comment]
> This LGTM. I tested on iOS and also Android since I noticed we did not test on those platforms. We should let QA know that they can test iOS and Android by clicking on the links with the app installed (at least this should work in theory).

Thanks a lot!

**Pull request merged**
